geode-core/src/main/java/org/apache/geode/cache/client/internal/QueueManagerImpl.java

```diff
@@ -862,8 +862,9 @@ public void recoverPrimary(Set<ServerLocation> excludedServers) {
       return;
     }
     final boolean isDebugEnabled = logger.isDebugEnabled();
-    if (queueConnections != null && queueConnections.getPrimary() != null
-        && !queueConnections.getPrimary().isDestroyed()) {
+    ConnectionList snapshot = queueConnections;
```

Comment:
I really don't like the name snapshot. It isn't a snapshot since the contents of the queue can change. It is the current reference to the queue. For volatiles I prefer either a local prefix, like `localQueueConnctions`, or just shadowing the member like:
```java
final ConnectionList queueConnections = this.queueConnections;
```
It can be made slightly more clear the intent with a little refactor.
```java
final ConnectionList queueConnections = getQueueConnections();
```

Reply:
Alternatively:
```java
if (primaryNotDestroyed(queueConnections)) {
   ...
}


...

static boolean primaryNotDestroyed(final ConnectionList connectionList) {
  if (null != connectionList) {
    final Connection primaryConnection = connectionList.getPrimary();
    if (null != primaryConnection) {
      return !primaryConnection.isDestroyed();
    }
  } 
  return false;
}
```
Now this new method is unit testable and guaranteed to be a non-volatile reference.